### PR TITLE
Enables easy use of local disks

### PR DESCRIPTION
**Pull request by ehankland**
Any benchmark that uses a scratch disk can now use a local disk by specifying --scratch_disk_type=local

### CODE REVIEW

**Comment by ehankland** [comment]
test this please.

**Comment by ehankland** [comment]
https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/issues/243

File: perfkitbenchmarker/aws/aws_disk.py [hunk 1]
```diff
@@ -33,6 +33,7 @@
 VOLUME_EXISTS_STATUSES = frozenset(['creating', 'available', 'in-use', 'error'])
 VOLUME_DELETED_STATUSES = frozenset(['deleting', 'deleted'])
 VOLUME_KNOWN_STATUSES = VOLUME_EXISTS_STATUSES | VOLUME_DELETED_STATUSES
+DISK_TYPE = {disk.STANDARD: 'standard', disk.SSD: 'gp2', disk.IOPS: 'io1'}
```
Comment (voellm) [inline perfkitbenchmarker/aws/aws_disk.py]: Should we rename IOPS to PIOPS since this is really not a general type?

I also think we should call SSD REMOTE_SSD to reduce confusion with LOCAL_SSD.

File: perfkitbenchmarker/aws/aws_virtual_machine.py [hunk 2]
```diff
@@ -143,6 +145,9 @@ def __init__(self, vm_spec):
     self.region = self.zone[:-1]
     self.image = self.image or GetImage(self.machine_type, self.region)
     self.user_name = FLAGS.aws_user_name
+    if self.machine_type in NUM_LOCAL_VOLUMES:
```
Comment (voellm) [inline perfkitbenchmarker/aws/aws_virtual_machine.py]: If we support multiple local disks should we also support striping?

Seems like most local disks will be striped with mdadm

File: perfkitbenchmarker/disk.py [hunk 3]
```diff
@@ -21,6 +21,11 @@
 
 from perfkitbenchmarker import resource
 
+STANDARD = 'standard'
+SSD = 'ssd'
+IOPS = 'iops'  # Provisioned IOPS (ssd) in AWS
```
Comment (voellm) [inline perfkitbenchmarker/disk.py]: Same comment as before... we should just call this PIOPS

File: perfkitbenchmarker/disk.py [hunk 4]
```diff
@@ -21,6 +21,11 @@
 
 from perfkitbenchmarker import resource
 
+STANDARD = 'standard'
+SSD = 'ssd'
```
Comment (voellm) [inline perfkitbenchmarker/disk.py]: Should call this REMOTE_SSD to be really clear what it is.

File: perfkitbenchmarker/disk.py [hunk 5]
```diff
@@ -21,6 +21,11 @@
...
+IOPS = 'iops'  # Provisioned IOPS (ssd) in AWS
+LOCAL = 'local'
```
Comment (voellm) [inline perfkitbenchmarker/disk.py]: LOCAL_SSD or LOCAL_ROTATIONAL_DISK ?

**Comment by voellm** [comment]
Just reviewed for functionality vs correctness.  It still needs a review.

**Comment by hungvelo** [comment]
Do we need to specify where this local disk is? Does this mean there will be some network traffic between controller machine (where scratch disk is located) and vm instance?

File: perfkitbenchmarker/aws/aws_disk.py (hunk 1 above)
Comment (ehankland) [inline perfkitbenchmarker/aws/aws_disk.py]: Renamed IOPS

File: perfkitbenchmarker/aws/aws_virtual_machine.py (hunk 2 above)
Comment (ehankland) [inline perfkitbenchmarker/aws/aws_virtual_machine.py]: I intentionally left striping the drives for another PR - if you want them in this one, that won't be too hard. What kind of flag do you want for that? Should we introduce a new one like --stripe_local_disks=True, or reuse --scratch_disk_type=striped_local, or just assume that benchmarks always want all the local drives striped?

File: perfkitbenchmarker/disk.py (hunk 4 above)
Comment (ehankland) [inline perfkitbenchmarker/disk.py]: Do you want the constant to be named REMOTE_SSD, the value (that people have to type with --scratch_disk_type) to be remote_ssd, or both?

File: perfkitbenchmarker/disk.py (hunk 3 above)
Comment (ehankland) [inline perfkitbenchmarker/disk.py]: Done

File: perfkitbenchmarker/disk.py (hunk 5 above)
Comment (ehankland) [inline perfkitbenchmarker/disk.py]: From PKB's point of view, this is very hard to tell the difference unless we start keeping track of which machine types have which kind of local drive (or check at run time whether the device says it's rotational or not). The intention was that local means whatever local drive you happen to have, ssd or rotational. Did you have an idea about how you wanted it to behave if someone specified --scratch_disk_type=local_ssd and the vm only had a rotational disk?

**Comment by ehankland** [comment]
"Local" disk here means local to the VM - most clouds have VM types that offer some kind of ephemeral storage (as opposed to a remote persistent store) bundled in with them - that's what this is.

**Comment by ehankland** [comment]
@voellm PTAL - had some questions

File: perfkitbenchmarker/disk.py (hunk 5 above)
Comment (ehankland) [inline perfkitbenchmarker/disk.py]: Talked offline - going to skip the separate local names - renaming ssd to remote_ssd is sufficient

File: perfkitbenchmarker/aws/aws_virtual_machine.py (hunk 2 above)
Comment (ehankland) [inline perfkitbenchmarker/aws/aws_virtual_machine.py]: Talked offline - will do this in a separate PR

File: perfkitbenchmarker/disk.py [hunk 6]
```diff
@@ -21,6 +21,11 @@
 
 from perfkitbenchmarker import resource
 
+STANDARD = 'standard'
+REMOTE_SSD = 'remote_ssd'
+PIOPS = 'piops'  # Provisioned IOPS (ssd) in AWS
```
Comment (cmccoy) [inline perfkitbenchmarker/disk.py]: Nit: Capitalize (SSD)

File: perfkitbenchmarker/virtual_machine.py [hunk 7]
```diff
@@ -122,6 +122,7 @@ def __init__(self, vm_spec):
     self.disk_specs = []
     self.scratch_disks = []
     self.hostname = None
+    self.max_local_drives = 0
```
Comment (cmccoy) [inline perfkitbenchmarker/virtual_machine.py]: Could you document this attribute?

File: perfkitbenchmarker/azure/azure_disk.py [hunk 8]
```diff
@@ -123,4 +125,7 @@ def Detach(self):
 
   def GetDevicePath(self):
     """Returns the path to the device inside the VM."""
-    return '/dev/sd%s' % chr(ord(DRIVE_START_LETTER) + self.lun)
+    if self.disk_type == disk.LOCAL:
```
Comment (cmccoy) [inline perfkitbenchmarker/azure/azure_disk.py]: Is this condition possible? `disk.LOCAL` isn't specified in `DISK_TYPE` above, and there's an assert in `_Create` that the `disk_type` is in `DISK_TYPE`.

Comment (ehankland) [inline perfkitbenchmarker/azure/azure_disk.py]: _Create doesn't get called for local disks - see CreateScratchDisk()

File: perfkitbenchmarker/virtual_machine.py (hunk 7 above)
Comment (ehankland) [inline perfkitbenchmarker/virtual_machine.py]: Done.

File: perfkitbenchmarker/disk.py (hunk 6 above)
Comment (ehankland) [inline perfkitbenchmarker/disk.py]: Done.

File: perfkitbenchmarker/azure/azure_disk.py (hunk 8 above)
Comment (cmccoy) [inline perfkitbenchmarker/azure/azure_disk.py]: OK, thanks.

**Comment by cmccoy** [comment]
LGTM. This has a few breaking changes that we need to be sure to publicize (`ssd` -> `remote_ssd`, dropped `--use_local_disk`). Maybe we should start a draft changelog for v0.16.0?

**Comment by ehankland** [comment]
Sure - as soon as v.15 goes in I'll update the changelog for v.16
